### ISSUE TITLE
Sclang Without X - need to rewrite X11 detection/interaction thanks to Qt5 

### ISSUE THREAD

**Issue by smrg-lm** [supercollider/supercollider#1736]
Hi

I remember that to run sclang without X server was a request that Jakob made available some time ago after the Qt integration. Now, when I try to run sclang in a tty I get:

QXcbConnection: Could not connect to display

That means that something has changed... is it possible to get back to the previous behavior?

Best
Lucas

**Comment by olafklingt** [comment]
you are right in the moment running sclang without x is not possible. And Even when compile without QtGui, you have to comment out the initialization of QtGUI in QtGUI.sc . and on my computer sclang runs with 100% processor usage so something must go wrong. I still have no idea where to start debugging.

**Comment by miczac** [comment]
besides that it would be great to have this sorted – as a temporary solution – could this help?
https://en.wikipedia.org/wiki/Xvfb

**Comment by danstowell** [comment]
I've found the problem:

Qt5 no longer provides the `QT_WS_X11` define. The consequence of this is that even on X11 systems, our build process had silently stopped activating the X11-specific code branches.

If we change the cmake code to "manually" provide a define, then we come up against a second problem: Qt5 no longer provides `QX11Info`, so the code in "hacks_x11" no longer compiles. It looks like it needs to be rewritten using "Qt Platform Abstraction" (according to [this thread](https://forum.qt.io/topic/19859/qt5-qx11info-display)) but I can't find any useful documentation about this new way.

So in other words, in order to fix sclang-without-x we need to fix sclang-with-x!

**Comment by danstowell** [comment]
Some attempts at this are in this branch here: https://github.com/danstowell/supercollider/tree/qt_without_x

**Comment by timblechmann** [comment]
setting the environment variable `QT_QPA_PLATFORM_PLUGIN=offscreen` might help?

**Comment by miczac** [comment]
Tried with QT_QPA_PLATFORM_PLUGIN set on Jessie on a R-Pi and the pre-packaged 3.6.6:
alas, doesn't change anything.
```
xvfb-run --auto-servernum /usr/bin/sclang "$ @" 
```
worked ...
(besides, then #1450 strikes again ... but that's something different)

**Comment by vivid-synth** [comment]
I'm able to run sclang in a tty now -- can anyone repro the above error?

**Comment by vivid-synth** [comment]
@smrg-lm are you able to repro this now?

**Comment by smrg-lm** [comment]
@vivid-synth I'm not in the latest git but in 0382c73d91166ff still happens.

**Comment by bagong** [comment]
Nothing relevant changed. You cannot run sclang with inbuilt qt in x-less/displayless terminals. Dan's analysis must be right. The loss of Q_WS_X11 has multiple bad side effects...

**Comment by bagong** [comment]
This simplest workaround seems to be to `export DISPLAY=:0` before you run sclang in an x-less environment. xvfb or other alternatives aren't really required.

**Comment by Apteryks** [comment]
This also affects tests when run inside a container (using Guix):

```
      Start 14: sclang_crash_1_run
14/17 Test #14: sclang_crash_1_run .................***Exception: Other  0.25 sec
/tmp/guix-build-supercollider-3.9.1.drv-0/build/lang/sclang: Relink `/gnu/store/b9ww6qv1ii9v6n45kin7543vkf6jfnd3-libpng-1.6.29/lib/libpng16.so.16' with `/gnu/store/n6nvxlk2j8ysffjh3jphn1k5silnakh6-glibc-2.25/lib/libpthread.so.0' for IFUNC symbol `longjmp'
/tmp/guix-build-supercollider-3.9.1.drv-0/build/lang/sclang: Relink `/gnu/store/1jp44pfmqcj3zycclvmmva3xcwinyg7l-freetype-2.8/lib/libfreetype.so.6' with `/gnu/store/n6nvxlk2j8ysffjh3jphn1k5silnakh6-glibc-2.25/lib/libpthread.so.0' for IFUNC symbol `longjmp'
SC FFT global init: cosTable initialised.
compiling class library...
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/guix-build-supercollider-3.9.1.drv-0/runtime-nixbld'
qt.qpa.screen: QXcbConnection: Could not connect to display :0
Could not connect to any X display.
terminate called without an active exception

      Start 15: sclang_crash_2_run
15/17 Test #15: sclang_crash_2_run .................***Exception: Other  0.21 sec
/tmp/guix-build-supercollider-3.9.1.drv-0/build/lang/sclang: Relink `/gnu/store/b9ww6qv1ii9v6n45kin7543vkf6jfnd3-libpng-1.6.29/lib/libpng16.so.16' with `/gnu/store/n6nvxlk2j8ysffjh3jphn1k5silnakh6-glibc-2.25/lib/libpthread.so.0' for IFUNC symbol `longjmp'
/tmp/guix-build-supercollider-3.9.1.drv-0/build/lang/sclang: Relink `/gnu/store/1jp44pfmqcj3zycclvmmva3xcwinyg7l-freetype-2.8/lib/libfreetype.so.6' with `/gnu/store/n6nvxlk2j8ysffjh3jphn1k5silnakh6-glibc-2.25/lib/libpthread.so.0' for IFUNC symbol `longjmp'
SC FFT global init: cosTable initialised.
compiling class library...
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/guix-build-supercollider-3.9.1.drv-0/runtime-nixbld'
qt.qpa.screen: QXcbConnection: Could not connect to display :0
Could not connect to any X display.
terminate called without an active exception

      Start 16: sclang_crash_3_run
16/17 Test #16: sclang_crash_3_run .................***Exception: Other  0.21 sec
/tmp/guix-build-supercollider-3.9.1.drv-0/build/lang/sclang: Relink `/gnu/store/b9ww6qv1ii9v6n45kin7543vkf6jfnd3-libpng-1.6.29/lib/libpng16.so.16' with `/gnu/store/n6nvxlk2j8ysffjh3jphn1k5silnakh6-glibc-2.25/lib/libpthread.so.0' for IFUNC symbol `longjmp'
/tmp/guix-build-supercollider-3.9.1.drv-0/build/lang/sclang: Relink `/gnu/store/1jp44pfmqcj3zycclvmmva3xcwinyg7l-freetype-2.8/lib/libfreetype.so.6' with `/gnu/store/n6nvxlk2j8ysffjh3jphn1k5silnakh6-glibc-2.25/lib/libpthread.so.0' for IFUNC symbol `longjmp'
SC FFT global init: cosTable initialised.
compiling class library...
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/guix-build-supercollider-3.9.1.drv-0/runtime-nixbld'
qt.qpa.screen: QXcbConnection: Could not connect to display :0
Could not connect to any X display.
terminate called without an active exception

      Start 17: sclang_crash_deprecated_run
17/17 Test #17: sclang_crash_deprecated_run ........***Exception: Other  0.21 sec
/tmp/guix-build-supercollider-3.9.1.drv-0/build/lang/sclang: Relink `/gnu/store/b9ww6qv1ii9v6n45kin7543vkf6jfnd3-libpng-1.6.29/lib/libpng16.so.16' with `/gnu/store/n6nvxlk2j8ysffjh3jphn1k5silnakh6-glibc-2.25/lib/libpthread.so.0' for IFUNC symbol `longjmp'
/tmp/guix-build-supercollider-3.9.1.drv-0/build/lang/sclang: Relink `/gnu/store/1jp44pfmqcj3zycclvmmva3xcwinyg7l-freetype-2.8/lib/libfreetype.so.6' with `/gnu/store/n6nvxlk2j8ysffjh3jphn1k5silnakh6-glibc-2.25/lib/libpthread.so.0' for IFUNC symbol `longjmp'
SC FFT global init: cosTable initialised.
compiling class library...
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/guix-build-supercollider-3.9.1.drv-0/runtime-nixbld'
qt.qpa.screen: QXcbConnection: Could not connect to display :0
Could not connect to any X display.
terminate called without an active exception


76% tests passed, 4 tests failed out of 17

Total Test time (real) =   8.53 sec

The following tests FAILED:
         14 - sclang_crash_1_run (OTHER_FAULT)
         15 - sclang_crash_2_run (OTHER_FAULT)
         16 - sclang_crash_3_run (OTHER_FAULT)
         17 - sclang_crash_deprecated_run (OTHER_FAULT)
Errors while running CTest
```

**Comment by brianlheim** [comment]
@Apteryks - you should be able to get around that with the approach here: https://github.com/supercollider/supercollider/blob/develop/testsuite/sclang/launch_test.py#L35-L39

**Comment by Apteryks** [comment]
@brianlheim thanks for the tip! I tried it, but it seems to take a long time (or simply hang) on teh first test requiring X (sclang_crash_1.scd_run). Should I wait more or is this not normal?

**Comment by brianlheim** [comment]
I'm not sure. @scztt may be able to answer better since he wrote the original script IIUC

**Comment by Apteryks** [comment]
OK, thanks for the reply. Definitely not normal: ............***Timeout 1500.01 sec :-). The commands I used were:
```
export DISPLAY=:99
export QT_PLATFORM_PLUGIN=offscreen
Xvfb :99 -ac -screen 0 1280x1024x16 &
make test
```

**Comment by scztt** [comment]
This trick was cribbed from some unknown stackoverflow, and has only been tested on Travis. :)
If you can come up with a more stable way to run UI tests on headless linux, it would be great to improve on the current workflow.

**Comment by dyfer** [comment]
I think this can be closed with the posted workaround:

> This simplest workaround seems to be to `export DISPLAY=:0` before you run sclang in an x-less environment. xvfb or other alternatives aren't really required.

**Comment by Apteryks** [comment]
@dyfer based on what I wrote in https://github.com/supercollider/supercollider/issues/1736#issuecomment-372053040, I don't think `export DISPLAY=:0` would suffice.

**Comment by dyfer** [comment]
Got it, thanks for clarifying!

**Comment by danstowell** [comment]
@dyfer I believe the original goal was any solution to run sclang without X server: the problem back in 2015 was that this possibility had been lost, because Qt threw an error. The best solution would be that sclang can be told to run without X, *even without recompiling sclang* - it should not invoke any X-dependent code/lib. An inferior solution would be an option based on recompiling sclang.

**Comment by dyfer** [comment]
@danstowell thanks for more info.

I checked a few things and I believe there are a few options here.

#### Running headless on systems with _no_ X installed 
Solution: **Build SC without Qt**

In order to build SC with Qt, X needs to be installed, so if X is not installed, building without Qt seems to be the only option. 

When installing SC from a package manager, Qt and X would be pulled in as dependencies I believe. I don't think it's feasible to request SC built with Qt support to "not load X11 libraries" - it would need to load these libraries at runtime and that's likely too cumbersome when building with Qt.

#### Running headless on systems with X installed

(This assumes that SC was built with Qt)
- option 1: use the qt `offscreen` plugin: `QT_QPA_PLATFORM=offscreen sclang`
  - in case of the `offscreen` plugin trying to create gui elements posts warnings
  - my guess is that this is the "lightest" approach, but that's only a guess
- option 2: use the qt `vnc` plugin: `QT_QPA_PLATFORM=vnc sclang`
  - in this case Qt creates a virtual display available / vnc server
  - this should allow to connect to that to view the GUIs; I have not tested this
- option 3: use `Xvfb` (X11 virtual framebuffer) as a display: ` xvfb-run sclang` (this may need additional parameters)
  - [this is what we use in our (headless) test environment](https://github.com/supercollider/supercollider/blob/develop/.github/workflows/actions.yml#L640-L642)

Please see if these options can be accepted as solutions to the original issue. If so, we should probably document them somewhere and (maybe?) close this issue.

**Comment by dyfer** [comment]
Actually, the using the `offscreen` plugin is [described in the Linux readme](https://github.com/supercollider/supercollider/blob/develop/README_LINUX.md#headless-operation). 

IMO we do have a solution for this and this issue can be closed.

**Comment by Apteryks** [comment]
@dyfer OK, sounds reasonable to close then! Thanks for looking at it!
